### PR TITLE
BUG: DataFrame constructor reordering elements with ndarray from datetime dtype not datetime64[ns]

### LINKED ISSUE #39422: BUG: pd.DataFrame is created inconsistently from numpy.ndarray for non-native pandas dtypes such as: datetime64[ms], etc..
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
import numpy as np

a = np.array(
    [
        ["2015-01-01", "2015-01-02", "2015-01-03"],
        ["2017-01-01", "2017-01-02", "2017-02-03"],
    ],
    dtype="datetime64[ms]",
)

print(pd.DataFrame(a))
           0          1          2
0 2015-01-01 2015-01-03 2017-01-02
1 2015-01-02 2017-01-01 2017-02-03


b = np.array(
    [
        ["2015-01-01", "2015-01-02", "2015-01-03"],
        ["2017-01-01", "2017-01-02", "2017-02-03"],
    ],
    dtype="datetime64[M]",
)

print(pd.DataFrame(b))
           0          1          2
0 2015-01-01 2015-01-03 2017-01-02
1 2015-01-02 2017-01-01 2017-02-03

c = np.array(
    [
        ["2015-01-01", "2015-01-02", "2015-01-03"],
        ["2017-01-01", "2017-01-02", "2017-02-03"],
    ],
    dtype="datetime64[ns]",
)

print(pd.DataFrame(c))
           0          1          2
0 2015-01-01 2015-01-02 2015-01-03
1 2017-01-01 2017-01-02 2017-02-03

```

#### Problem description
The exists an inconsistency how pd.DataFrame() is created from 2D numpy.ndarray for not pandas native type. The dataframe should be transposed for datetime64[M], datetime64[ms] etc.

#### Expected Output
```python
print(pd.DataFrame(a))
           0          1          2
0 2015-01-01 2015-01-02 2015-01-03
1 2017-01-01 2017-01-02 2017-02-03

print(pd.DataFrame(b))
0 2015-01-01 2015-01-02 2015-01-03
1 2017-01-01 2017-01-02 2017-02-03
```

#### Possible Hint
```python

d = np.array(
    [
        ["2015-01-01", "2015-01-02", "2015-01-03"],
        ["2017-01-01", "2017-01-02", "2017-02-03"],
    ],
    dtype="datetime64[ms]",
    order="f",
)

print(pd.DataFrame(d))
0 2015-01-01 2015-01-02 2015-01-03
1 2017-01-01 2017-01-02 2017-02-03

```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : db08276bc116c438d3fdee492026f8223584c477
python           : 3.8.5.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United Kingdom.1252

pandas           : 1.1.3
numpy            : 1.19.2
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 21.0
setuptools       : 49.6.0.post20210108
Cython           : None
pytest           : 6.2.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.3
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.0
sqlalchemy       : 1.3.22
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #39422
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

The ravel("K") was a problem when passing in a view on the underlying object, which was done here.

### CODE REVIEW

File: pandas/_libs/tslibs/conversion.pyx [hunk 1]
```diff
@@ -222,7 +222,7 @@ def ensure_datetime64ns(arr: ndarray, copy: bool=True):
         dtype = arr.dtype
         arr = arr.astype(dtype.newbyteorder("<"))
 
-    ivalues = arr.view(np.int64).ravel("K")
+    ivalues = arr.view(np.int64).ravel("C")
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/conversion.pyx:225]: this is going to make a copy if arr has order="F" isnt it?

File: pandas/tests/frame/test_constructors.py [hunk 2]
```diff
@@ -1762,6 +1762,37 @@ def test_constructor_datetimes_with_nulls(self, arr):
...
+        ],
+    )
+    def test_constructor_datetimes_non_ns(self, dtype):
+        na = np.array(
+            [
+                ["2015-01-01", "2015-01-02", "2015-01-03"],
+                ["2017-01-01", "2017-01-02", "2017-02-03"],
+            ],
+            dtype=dtype,
+        )
```
Comment (jbrockmendel) [inline pandas/tests/frame/test_constructors.py:1785]: can you also parametrize over contiguity

File: pandas/_libs/tslibs/conversion.pyx (hunk 1 above)
Comment (phofl) [reply pandas/_libs/tslibs/conversion.pyx:225]: Yeah unfortunately, but not sure how we can avoid this here. Could transpose back before passing in, but I think that is not an ideal solution

File: pandas/tests/frame/test_constructors.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/frame/test_constructors.py:1785]: Like this?

File: pandas/_libs/tslibs/conversion.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/conversion.pyx:225]: could arr.flat be useful here?

File: pandas/tests/frame/test_constructors.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/frame/test_constructors.py:1787]: yes, thanks

File: pandas/_libs/tslibs/conversion.pyx [hunk 3]
```diff
@@ -222,7 +222,7 @@ def ensure_datetime64ns(arr: ndarray, copy: bool=True):
...
 
     result = np.empty(shape, dtype=DT64NS_DTYPE)
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/conversion.pyx:227]: would it make a difference it we constructed this with `result = np.empty_like(arr, dtype=...)` to ensure we keep the same contiguity?

Comment (phofl) [reply pandas/_libs/tslibs/conversion.pyx:227]: Oh yes, fantastic. Thanks very much

File: pandas/tests/frame/test_constructors.py [hunk 4]
```diff
@@ -1762,6 +1762,39 @@ def test_constructor_datetimes_with_nulls(self, arr):
         expected = Series([np.dtype("datetime64[ns]")])
         tm.assert_series_equal(result, expected)
 
+    @pytest.mark.parametrize("order", ["K", "A", "C", "F"])
```
Comment (jreback) [inline pandas/tests/frame/test_constructors.py:1765]: can you do the same with timedelta? not sure if that is broken as well.

Comment (phofl) [reply pandas/tests/frame/test_constructors.py:1765]: This uses numpy astype under the hood, is fine. Added tests. 

Is there a way to create a df without a numpy or pandas array which has a dtype other than timedelta64[ns]?
I casted back to compare them

Comment (jreback) [reply pandas/tests/frame/test_constructors.py:1765]: umm not sure i understand what you are asking

Comment (phofl) [reply pandas/tests/frame/test_constructors.py:1765]: Sorry, here an example:
```
expected = DataFrame(
    [
        [Timedelta(days=1), Timedelta(days=2)],
        [Timedelta(days=3), Timedelta(days=4)]
    ], dtype="timedelta64[ms]"
)
```

returns 
```
             0            1
0   86400000.0  172800000.0
1  259200000.0  345600000.0
```
with dtype float
while

```
na = np.array(
    [
        [np.timedelta64(1, 'D'), np.timedelta64(2, 'D')],
        [np.timedelta64(4, 'D'), np.timedelta64(5, 'D')]
    ],
    dtype="timedelta64[ms]",
)
df = DataFrame(na)
```
returns 
```
       0      1
0 1 days 2 days
1 4 days 5 days
```

where dtype is timedelta64[ms]

Comment (phofl) [reply pandas/tests/frame/test_constructors.py:1765]: Question is if we can create the expected in a way to have timedelta64[ms] without using numpy or pandas array

Comment (jbrockmendel) [reply pandas/tests/frame/test_constructors.py:1765]: i dont think so, no.  if it were float or int we could use a memoryview directly

Comment (phofl) [reply pandas/tests/frame/test_constructors.py:1765]: Casting to timedelta64[ns] may seem a bit odd, but does what we want here, so I would keep the current layout. Thx

Comment (jreback) [reply pandas/tests/frame/test_constructors.py:1765]: yeah agree, i think what you did is fine for the tests.

**Comment by jreback** [comment]
thanks @phofl

**Comment by jorisvandenbossche** [comment]
This was a regression, so I think worth backporting to 1.2.2. If we do that, we only need a small PR moving the whatsnew note to v1.2.2.rst (apart from the actual backport)

**Comment by jreback** [comment]
sure we could backport

@phofl if u wouldn't mind putting up a backport PR (need to edit the note)

and then separate PR to move the note in master

**Comment by phofl** [comment]
Done #39475 and #39476

**Comment by jorisvandenbossche** [comment]
Thanks @phofl !

**Pull request merged**